stories/2024-01-01-Eurovision/story.ipynb

```diff
@@ -2840,6 +2840,674 @@
+                "\n",
+                "    # violin plot of predictions for each rank\n",
+                "    sns.violinplot(x=\"actual\", y=\"predictions\", data=prediction_df, order=range(0, 13), ax=ax1)\n",
+                "    sns.violinplot(y=\"actual\", x=\"predictions\", data=prediction_df, order=range(0, 13), ax=ax2)\n",
+                "\n",
+                "    # add a best fit line - scale is wrong because numeric on top of categorical\n",
+                "    sns.regplot(x=\"actual\", y=\"predictions\", data=prediction_df, scatter=False, color='black', ax=ax1)\n",
+                "    sns.regplot(y=\"actual\", x=\"predictions\", data=prediction_df, scatter=False, color='black', ax=ax2)\n",
+                "\n",
+                "\n",
+                "    # spearman correlation\n",
+                "    corr, _ = spearmanr(prediction_df['actual'], prediction_df['predictions'])\n",
+                "\n",
+                "    fig.suptitle('Predicted Score vs Actual Score (r = ' + str(round(corr, 4)) + ')', fontsize=15)\n",
+                "\n",
+                "    # set x label\n",
+                "    ax1.set_xlabel('Actual Score')\n",
+                "    ax1.set_ylabel('Predicted Score')\n",
+                "    ax2.set_ylabel('Actual Score')\n",
+                "    ax2.set_xlabel('Predicted Score')\n",
+                "\n",
+                "def cm_heatmap (prediction_df, title='Confusion matrix', ax=None):\n",
+                "    cm = confusion_matrix(prediction_df['actual'], prediction_df['predictions'])\n",
+                "\n",
+                "    plot = sns.heatmap(cm, annot=True, fmt='g', ax=ax,\n",
+                "                    vmin=0, vmax=100, cmap='Blues')\n",
+                "\n",
+                "    if(len(cm[[0]][0]) == 11) :\n",
+                "    # change the tick labels\n",
+                "        plot.set_xticks(np.arange(0, 11, 1)+0.5, np.arange(0, 9, 1).tolist() + [10, 12])\n",
+                "        plot.set_yticks(np.arange(0, 11, 1)+0.5, np.arange(0, 9, 1).tolist() + [10, 12])\n",
+                "    else:\n",
+                "        plot.set_xticks(np.arange(0, 10, 1)+0.5, np.arange(1, 9, 1).tolist() + [10, 12])\n",
+                "        plot.set_yticks(np.arange(0, 10, 1)+0.5, np.arange(1, 9, 1).tolist() + [10, 12])\n",
+                "\n",
+                "    # add tick marks\n",
+                "    plot.tick_params(axis='both', which='both', length=5, color='black')\n",
+                "\n",
+                "    plot.set_title(title)\n",
+                "    plot.set_ylabel('Actual Score')\n",
+                "    plot.set_xlabel('Predicted Score')"
+            ]
+        },
+        {
+            "cell_type": "code",
+            "execution_count": null,
+            "metadata": {},
+            "outputs": [],
+            "source": [
+                "violins(out)"
```

Comment:
those are super helpful to understand what's going on. It seems like there is definitely some predictive quality to the model, though not very precise (except for the top score categories maybe)